---
title: "Differential Gene Expression Analysis using dream"
author: "Oumie Kuyateh"
date: "24 June 2023"
output: html_document

---

The aim of this analysis is to do Differential Expression (DE) analyses using dream. dream is built on limma/voom but for more than one random effect to be put in the model. We will use nora viruses, which is found in one fo the daatsets we analysed as an example of the expression analysis. 



Software Versions 

R version 4.2.0 (2022-04-22)
edgeR version 3.40.0
ggplot2 version 3.4.2
dplyr version 1.1.2 
tibble version 3.2.1
variancePartition version 1.28.0
BiocParallel version 1.32.5
ggfortify version 
AnnotationDbi version 1.60.0 
org.Dm.eg.db version 3.16.0 


Firstly, Remove everything files from R's environment


```{r}


rm(list = ls())
```


Now load nora_virus the libraries I will need
```{r}


library(edgeR)
library(ggplot2)
library(dplyr)
library(tibble) 
library(variancePartition)
library(BiocParallel)
library(AnnotationDbi)
library(org.Dm.eg.db)

```


Read in the file names of the count matrix files from featureCounts and store them in a vector called "Count_Matrix_Filenames"

```{r}


Count_Matrix_Filenames <- list.files("/mnt/drive5-6tb/Oumie/Dream_Analysis_Conda_Updated_packages", pattern="Processed_Count_Matrix")


```

Read in the featureCounts files and store them in a list called "Count_Matrix_File_List" and give them the names of the Count_Matrix_Filenames

```{r}

Count_Matrix_File_List <- lapply(Count_Matrix_Filenames, function(x) read.table(x, header = TRUE,sep = "\t", row.names = 1) )



names (Count_Matrix_File_List) <- Count_Matrix_Filenames 



```


Read in the file names of the metadata files of the datasets and store in a vector called "Metadata_Filenames  

```{r}


Metadata_Filenames <- list.files("/mnt/drive5-6tb/Oumie/Dream_Analysis_Conda_Updated_packages", pattern="Processed_Metadata")


```

Read in the metadata files and store them in a list called "Metadata_File_List" and give them the names of the Metadata_File_List

```{r}

Metadata_File_List <- lapply(Metadata_Filenames, function(x) read.table(x, header = TRUE,sep = "\t", row.names = 1) )

names (Metadata_File_List) <- Metadata_Filenames 

```


The Count matrices and metadata files have the date they were saved on the computer attach to their names, so we remove them using gsub

```{r}

names (Count_Matrix_File_List) <- gsub ("_19_06_2023.txt", "", names (Count_Matrix_File_List))

names (Metadata_File_List) <- gsub ("_19_06_2023.txt", "", names (Metadata_File_List))


```

Now put the count matrices and metadata in the lists "Count_Matrix_File_List" and "Metadata_File_List" repectively in the global environment

```{r}

list2env (Count_Matrix_File_List, envir = .GlobalEnv)

list2env (Metadata_File_List, envir = .GlobalEnv)
```

For this analysis of the effect of nora virus infection on gene expression, we will combine the count matrices of the five datasets cwith detectable levels of nora virus infection using using bind_rows from the dplyr package.These daatsets are "PRJNA483441", " PRJNA518903", "PRJNA325107", "PRJNA281652", "PRJNA261333". Put allt hese in a list called "nora_virus_Count_Matrix_df"

```{r}
nora_virus_Count_Matrix_df <-  dplyr::bind_rows (Processed_Count_Matrix_PRJNA483441, Processed_Count_Matrix_PRJNA518903, Processed_Count_Matrix_PRJNA325107, Processed_Count_Matrix_PRJNA281652, Processed_Count_Matrix_PRJNA261333)


#Since some of the genes did not have any registered count in some datasets (i.e. genes with a count of zero in all samples in each dataset were removed whilst processing the count data from featurecounts an will be given a value of NA after using the "dplyr::bind_rows" function earlier. So we now replace the NAs to 0)


nora_virus_Count_Matrix_df [is.na(nora_virus_Count_Matrix_df)] <- 0


#Add a value of 1 to everything since a log base 2 will be used in the normalization 


nora_virus_Count_Matrix_df <- nora_virus_Count_Matrix_df + 1


```



Transverse the  "nora_virus_Count_Matrix_df" to make the genes as the genes and the libraries as the columns

```{r}
nora_virus_Count_Matrix_df <- as.data.frame (t(nora_virus_Count_Matrix_df))
```



The nora_virus_Count_Matrix_df data has FlyBase identifier (FBgn#) as the rownames. To get the GeneSymbol, we use data from the the "org.Dm.eg.db" AnnotationDb object which was loaded in the beginning of our analysis. 


```{r}
#To do this, we use the "mapIds()" function, to look in the rownames of the nora_virus_Count_Matrix_df  which contain the "FLYBASE" keytype. For each"FLYBASE"keytype return the Gene SYmbol. If multiple Symbols are available, give me the first one. Save this in a data.frame cnora_virused "FBgn_to_Symbol".


FBgn_to_Symbol <-  FBgn_to_Symbol[ order(row.names( FBgn_to_Symbol)),, drop =FALSE]

#First order the FBgn_to_Symbol and the nora_virus_Count_Matrix_df by rownames 

FBgn_to_Symbol <- data.frame( mapIds(org.Dm.eg.db, names (nora_virus_Count_Matrix_df ),  column="SYMBOL", keytype="FLYBASE", multiVals="filter"))


nora_virus_Count_Matrix_df <- nora_virus_Count_Matrix_df [ order(row.names( nora_virus_Count_Matrix_df )),, drop =FALSE]


#Check that they are in the same order


print (paste (" The row.names of FBgn_to_Symbol and nora_virus_Count_Matrix_df are in the same order:", identical(row.names(nora_virus_Count_Matrix_df ), row.names (FBgn_to_Symbol) )))



#Now attach a column in nora_virus_Count_Matrix_df containing the gene_Symbols


nora_virus_Count_Matrix_df$Gene_Symbols <- FBgn_to_Symbol[,1]


#Some FBgn numbers do not have corresponding gene Symbols and will be put as NA. However, we cannot have NA as ronames so remove them  


nora_virus_Count_Matrix_df <- na.omit(nora_virus_Count_Matrix_df )

#Now make the gene names as row.names but first remove the current row.names which is the FBgn numbers

row.names(nora_virus_Count_Matrix_df ) <- NULL


#Now make the gene symbols the rownames


nora_virus_Count_Matrix_df  <- column_to_rownames(nora_virus_Count_Matrix_df , var = "Gene_Symbols")



#Now export this combined Countmatrix and save it as "nora_virus_Count_Matrix_19_06_2023.txt"


write.table(nora_virus_Count_Matrix_df, file = "nora_virus_Count_Matrix_19_06_2023.txt",  sep = "\t")

```


The analysis of the effect of nora virus infection on gene expression, we will combine the metadata of the five datasets with detectable levels of nora virus infection using using bind_rows from the dplyr package.These daatsets are "PRJNA483441", " PRJNA518903", "PRJNA325107", "PRJNA281652", "PRJNA261333". Put allt hese in a list called "nora_virus_Metadata_df"
```{r}

#For this analysis of Sex differences in all the datasets, we will combine the metadata using bind_rows from the dplyr package

nora_virus_Metadata_df <- dplyr::bind_rows ( Processed_Metadata_PRJNA483441_Everret, Processed_Metadata_PRJNA518903_Maria, Processed_Metadata_PRJNA325107_Wen, Processed_Metadata_PRJNA281652_Kurmangaliyev, Processed_Metadata_PRJNA261333_Lee)



#Since there is a variance in the metadata variables that are present between the datasets, we will remove the columns with NA )



nora_virus_Metadata_df <- nora_virus_Metadata_df [ , colSums(is.na(nora_virus_Metadata_df ))==0]



#Now make the Females as the baseline in the analysis using factor function


nora_virus_Metadata_df$featureCounts_Sex <- factor(nora_virus_Metadata_df$featureCounts_Sex, levels = c("Female", "Male"))

#Check that female is the baseline


unique (nora_virus_Metadata_df$featureCounts_Sex)

#Look at the column nmaes of the Metadata to make sure all the required variables for the model are in 


names (nora_virus_Metadata_df)


#Now export this combined Metadata and save it as "nora_virus_Metadata_19_06_2023.txt"


write.table(nora_virus_Metadata_df  , file = "nora_virus_Metadata_19_06_2023.txt",  sep = "\t")
```


The columnames of the nora_virus_Count_Matrix_df should be the same as the rownames of the nora_virus_Metadata_df i.e. they must have the same library or sample name. However, we need to test that they are identical as a QC step. To do this, we first order the nora_virus_Metadata_df   and the nora_virus_Count_Matrix_df  by rownames and columnnames respectively.


```{r}

#First order the rownames of the nora_virus_Metadata_df 


nora_virus_Metadata_df  <- nora_virus_Metadata_df   [order(row.names(nora_virus_Metadata_df  )),]


#Then order the column names of the nora_virus_Count_Matrix_df


nora_virus_Count_Matrix_df  <- nora_virus_Count_Matrix_df [ , order(colnames(nora_virus_Count_Matrix_df ))]



#Now check that the rownames of the nora_virus_Metadata_df   and colnames of nora_virus_Count_Matrix_df  are in the same order



print (paste ("row.names of nora_virus_Metadata_df   and colnames of nora_virus_Count_Matrix_df  are identical:", identical(row.names( nora_virus_Metadata_df  ), colnames (nora_virus_Count_Matrix_df ) )))


```


Now we move the expression analysis. First Filter genes by number of counts in the nora_virus_Metadata_df   data frame  by Summing the counts per million (cpm) that are above 0.1 for each gene and then select the genes whose sum is greater than or equal to 5 cpm. Save these genes in isexpr
```{r}


isexpr <- rowSums(cpm(nora_virus_Count_Matrix_df ) > 0.1) >= 5


```


Now create a DGEList object from the nora_virus_Count_Matrix_df  of nora_virus the genes in isexpr that are in nora_virus_Count_Matrix_df  and save this as "geneExpr.

```{r}

geneExpr <- DGEList( nora_virus_Count_Matrix_df [isexpr,] )



```


Now calculate normalization facotors from the "geneExpr" DGEList object using calcNormFactors. NB calcNormFactors doesn't normalize the data, it just calculates normalization factors for use downstream.

```{r}

geneExpr <-calcNormFactors( geneExpr )
```

The dream method replaces two core functions of limma with a linear mixed model. As mentioned earlier, voomWithDreamWeights  is the same as what described above for voom(), except that it allows random effects in the formula.


Specify parnora_virusel processing parameters. This is used implicitly by dream() to run in parallel. It allows three types of clusters to be used. The computer has 24 cores, so we will use 23 of them in the analysis


```{r}


param <- SnowParam(23, "SOCK", progressbar=TRUE)


register(param)



```


Specify the model to be tested and save this as form. We are using "featureCounts_Sex" as it is the sex we confirmed based on the number of reads that mapped to sex-specific genes.

```{r}


form <- ~ featureCounts_Sex +  nora_virus_Presence + nora_virus_Presence:featureCounts_Sex + (1|All_Random_Effects) 


```


Estimate weights using linear mixed model of dream

```{r}

vobjDream <- voomWithDreamWeights ( geneExpr, form, nora_virus_Metadata_df   )
```

Fit the dream model on each gene. By default, uses the Satterthwaite approximation for the hypothesis test

```{r}

fitmm <- dream( vobjDream, form, nora_virus_Metadata_df  )
```

Now look at the coefficients of the model

```{r}

head(fitmm$coefficients)
```


Now save the coefficients in daatframes for to export later
```{r}
#Save the "Intercept" coefficient in a dataframe called Intercept 


Intercept<- topTable( fitmm , coef="(Intercept)", n = Inf )




#Save the "featureCounts_SexMale" coefficient in a dataframe called Sex_Effect 


Sex_Effect <- topTable( fitmm , coef="featureCounts_SexMale", n = Inf )




#Save "nora_virus_PresenceYes" coefficient in a dataframe called Sex_Effect 


nora_virus_Effect  <- topTable( fitmm, coef="nora_virus_PresenceYes", n = Inf )




#Now put the "featureCounts_SexMale:nora_virus_PresenceYes" coefficient in a dataframe called Sex_Effect 


nora_virus_and_Sex_Interaction_Effect  <- topTable( fitmm, coef="featureCounts_SexMale:nora_virus_PresenceYes", n = Inf )



```


```{r}

#Export the results for each coefficient extracted above as a txt file.

write.table(Intercept, file = "Intercept_nora_virus_Dream_24_06_2023.txt", sep = "\t")

write.table(Sex_Effect, file = "Sex_Effect_Only_nora_virus_Dream_24_06_2023.txt", sep = "\t")

write.table(nora_virus_Effect, file = "nora_virus_Effect_Dream_24_06_2023.txt", sep = "\t")

write.table(nora_virus_and_Sex_Interaction_Effect, file = "nora_virus_Sex_Interaction_Dream_24_06_2023.txt", sep = "\t")
```

